Safe versions of evil-join and comment-or-uncomment

yuhan0 opened the pull request:
Here are two evil operators I've written that make regular commenting and joining operations behave in a way that preserves balanced lisp structure, like the rest of lispyville. :)
These gifs should do a good job of explaining what they do:
![join-operator](https://user-images.githubusercontent.com/22216124/48501862-31dc2d00-e879-11e8-8c89-867a85edef36.gif)
![comment-or-uncomment-operator](https://user-images.githubusercontent.com/22216124/48501859-30126980-e879-11e8-972b-fa6b8d57db0d.gif)


I realise there should be proper ERT tests written like for the rest of the commands, but still haven't figured out yet how to set them up..

noctuid commented:
Awesome! Thanks so much.

I'll add a CONTRIBUTING file with instructions on running/writing tests and will review your changes.

noctuid commented on lispyville.el:
```diff
@@ -1890,7 +1958,10 @@ When THEME is not given, `lispville-key-theme' will be used instead."
-           [remap evil-change-whole-line] #'lispyville-change-whole-line))
+           [remap evil-change-whole-line] #'lispyville-change-whole-line
+           [remap evil-join] #'lispyville-join)
+         (lispyville--define-key 'normal
```
Remapping `evil-join` is fine, but I don't want to override a user's binding for `gc` by default. Please remove the keybinding or put it in a separate `comment` key theme.

noctuid commented on lispyville.el:
```diff
@@ -665,6 +665,74 @@ ARG has the same effect."
+         ;; must include trailing whitespace to avoid error with undo-tree
+         (progn (goto-char (cdr r))
+                (skip-chars-forward "[:space:]") (point)))))
+    (delete-trailing-whitespace beg (marker-position end-marker)))
```
`end-marker` should work fine by itself I believe. `marker-position` should not be necessary.

noctuid commented on lispyville.el:
```diff
@@ -665,6 +665,74 @@ ARG has the same effect."
+         (car r)
+         ;; must include trailing whitespace to avoid error with undo-tree
+         (progn (goto-char (cdr r))
+                (skip-chars-forward "[:space:]") (point)))))
```
Please move the call to `point` to its own line.

noctuid commented on lispyville.el:
```diff
@@ -665,6 +665,74 @@ ARG has the same effect."
+  (when (string-match-p (format "[%s]+$" (regexp-quote comment-start))
+                        (buffer-substring beg end))
+    (setq end (point-at-eol)))
+  (let* ((end-marker (copy-marker (1+ end))))
```
Please change to `let` since there is currently only one binding.

noctuid commented on lispyville.el:
```diff
@@ -665,6 +665,74 @@ ARG has the same effect."
+May insert newlines where necessary in order to move delimiters outside of the
+commented regions and preserve the balanced structure. "
+  (interactive "<R>")
+  (when (string-match-p (format "[%s]+$" (regexp-quote comment-start))
```
Could you explain the purpose of this check? Also, `comment-start` is not always a single character (e.g. in c-mode it is `/* `); I'm guessing `comment-start-skip` should be used instead.

yuhan0 replied:
This was due to a strange issue I encountered with undo-tree that would happen in the following situations:
- With the cursor in front of multiple comment characters, attempt to comment-or-uncomment only one character ("gcl")
```
|;; some comment
```
Result: the entire line is uncommented, but all previous undo history in the buffer gets completely mangled. 
Trying to `undo-tree-undo` produces the error:
```
primitive-undo: changes to function different than announced
```
- attempting to comment a word in the middle of a line:
```
foo |bar baz

;; "gce" ->

foo ;; bar
baz
```
but corrupts undo history as well, explaining the check slightly further down.

After spending way too long trying to dig into the root cause (appears to be an upstream bug with either `uncomment-region` or `undo-tree`) I ended up just adding the above as workarounds 😕

noctuid replied:
That's fine for now, but hopefully this can be fixed in `undo-tree` and removed in the future. Can you add a similar comment above this check too explaining that it's a workaround for `undo-tree`? I've stopped using `undo-tree` entirely because of corruption issues personally.

noctuid commented on lispyville.el:
```diff
@@ -665,6 +665,74 @@ ARG has the same effect."
+  (let ((inline-comment
+         (delete-and-extract-region
+          (save-excursion (forward-line 0)
+                          (search-forward comment-start (point-at-eol) 'no-error)
```
Same comment about `comment-start` here and below in this function. `comment-search-forward` uses `comment-start-skip` when searching and takes a limit as well. I'd suggest looking into using that instead.

noctuid commented on lispyville.el:
```diff
@@ -665,6 +665,74 @@ ARG has the same effect."
+          (point-at-eol))))
+    (join-line 1)
+    (indent-according-to-mode)
+    (if (search-forward comment-start (point-at-eol) 'no-error)
```
Use `when` instead of `if`.

noctuid commented on lispyville.el:
```diff
@@ -665,6 +665,74 @@ ARG has the same effect."
+           ((save-excursion (forward-line 1) (back-to-indentation) (lispy--in-comment-p))
+            (delete-region (point-at-eol)
+                           (progn (forward-line 1)
+                                  (skip-chars-forward (concat "[:space:]" (regexp-quote comment-start)))
```
Skip over `comment-start-skip` instead (e.g. use `looking-at` and `match-end` if it is succesful).

noctuid commented:
I've added a [contributing file](https://github.com/noctuid/lispyville/blob/master/CONTRIBUTING.org) with explanations on running/writing tests. Please let me know if you can't get the tests to run or have any questions. There are some annoyances with the current ERT setup, and I will be switching to buttercup in the future, but for now ERT is what lispyville uses.

Again, thanks for the PR; very cool additions.

yuhan0 commented:
Thanks for the suggestions :) 
I forgot to consider alternative multi-line comment syntaxes, it makes sense that one would want to use these commands outside of lisp modes. Will probably get around to it over the weekend!

yuhan0 commented:
I've made most of the requested changes, and added a few new ones that are still a little work-in-progress.  
The undo-tree bug turned out to be an issue with the new combine-change-calls mechanism in bleeding-edge Emacs 27, so temporarily setting the `comment-combine-change-calls` variable to nil removes the need for the previous checks. I also switched back to Emacs 26.1, might revisit this when an official version of 27 is released.  

Created a new key theme called `commentary` since that's where the "gc" binding comes from: <https://github.com/linktohack/evil-commentary>, also added a comment-or-uncomment-lines operator bound to "s-/".  

It also got me experimenting with a comment-and-yank operator bound to "gy" for trying out code "variations" - however dealing with unbalanced lines is a little less straightforward than with other operators - consider the case  

```elisp
(foo (bar|
      baz))
```

The regular operation (evil-commentary-copy-and-yank) `"gyyp"` preserves the structure, but the commented line(s) are unbalanced and will break the structure if uncommented - not too big of a deal, but a bit of pain when the other operators are all "safe"  - e.g. not being able to uncomment the 1st line and then delete the entire 2nd line afterwards  with "dd"

```elisp
;; (foo (bar
(foo (bar|
      baz))
```

Another possibility is stripping out the unbalanced parens when yanking (like the existing lispyville-yank), then inserting the safe yanked text as a comment  ``"yyP`[gc`]"``

```elisp
;; foo bar
(foo (bar|
      baz))
```

Or a more "dwim" approach which figures you don't *really* want to copy-and-comment the entire line, only the rightmost balanced region:  

```elisp
(foo (;; bar
      bar|
      baz))
```

I implemented the last two as an experiment and bound them to `gy` and `gY to see which would be  more useful, do let me know if you have any preferences :)

yuhan0 commented:
Also I took the liberty of writing a simple syntax transformation macro for the tests which allows for writing this 

```elisp
(should (string= (lispyville-with "((\n  ~|(a b)))"
                                  "D")
                 "|(())"))
```

in a more readable form with "before" and "after" strings vertically aligned:  

```elisp
(lispyville-should ("D")
  "((\n  ~|(a b)))"
  "|(())")
```

noctuid commented:
Thanks for adding the documentation. For consistency with the other source blocks, please use lowercase `(begin|end)_src` and remove the extra indentation (set `org-src-preserve-indentation t`). Also, change the example to use `evil-define-key` instead of `lispyville--define-key` (`lispyville--define-key` is an internal helper and has had breaking changes in the past).

As for `gy`, I don't know what the best behavior would be as I haven't really thought about doing this with an unbalanced region before. Normally, I would use `lispy-clone` and then `lispy-comment` to do this with an entire function or sexp. I think it might be better to have `gy` act like `lispyville-yank` by default for consistency, but I'll let you decide. If you think the second/dwim approach would be useful more often, feel free to give it the better keybinding. After you've decided on the behavior, please add tests for both. I haven't tried it yet, but does the version that works on the rightmost safe region work well with multi-line unbalanced regions?

Let me know when you're done making changes and think it's ready to merge.

noctuid commented:
> Also I took the liberty of writing a simple syntax transformation macro for the tests which allows for writing this

Thanks. I didn't put it in the contributing file yet, but if you have any more complicated examples, you can keep the literal newlines so things are more readable:
```elisp
(lispyville-should (...)
  "
|(foo
 (bar baz
      (qux ...)))"

  "
|(foo...")
```

yuhan0 commented:
Thanks for the feedback :) I used lispy-stringify and unstringify to toggle between multiline representations when writing the tests, so that wasn't too much of a problem.  

Regarding the `gy` operators, I agree it's a rather obscure scenario, mostly for the sake of laziness - using a sloppy "gyy" on the current line and having it dwim, rather than thinking and navigating to another location to execute structural commands.

>  I think it might be better to have `gy` act like `lispyville-yank` by default for consistency, but I'll let you decide.

Personally I'm leaning towards the second version since you'd normally use the command for making a "backup" of some state to be restored, and stripping away unbalanced parens kind of defeats the purpose. 
I wouldn't really want to keep both operators and bindings in the final version, how about a `C-u` prefix argument to toggle behavior?  

> I haven't tried it yet, but does the version that works on the rightmost safe region work well with multi-line unbalanced regions?

For any larger region than a line I'd use a balanced text object eg. `gyal` , but maybe mapping the copy-and-comment over all balanced sub-regions rather only the rightmost one makes more sense in that case:

![2018-11-26 02 42 56](https://user-images.githubusercontent.com/22216124/48983170-91162900-f126-11e8-8daf-9acdecc5ea4f.gif)

noctuid commented:
Sorry for the late response (meant to give one earlier but had problems with the github mobile app I'm using).

> Personally I'm leaning towards the second version since you'd normally use the command for making a "backup" of some state to be restored...

Given the intended use case, I think it would be fine to just keep the dwim version and remove the first version entirely. If someone actually wants it, it wouldn't be hard to add then. As long as intended use case for the dwim operator is documented (comment/experiment/restore), then it shouldn't be confusing. I would put `dwim` in the operator name though.

> maybe mapping the copy-and-comment over all balanced sub-regions 

I like this behavior, but I'm guessing the operator wouldn't be used often with regions like in the example (it probably wouldn't be useful). For now, I think it would be fine to only act on the rightmost region with the dwim operator unless you can think of use cases where this would be undesirable. Do whatever you think is best.

yuhan0 commented:
Alright, I changed the name of the operator to `lispyville-comment-and-copy-dwim` and added a few tests and documentation. Let me know if you need anything else :)

yuhan0 commented:
Come to think of it, maybe `comment-and-clone-dwim` would be a better name, since nothing is being added to the kill ring (unlike the evil-commentary version)

noctuid commented:
Sorry again for the late response. The norm is for operators/commands to be bound in normal state. The documentation saying normal state was the default was inconsistent with the implementation and example code; I went ahead and fixed this myself.

Also the tests do not all pass locally for me (specifically the `lispyville-comment-and-clone-dwim` test with `gyy` and the lorem ipsum text fails). It's not clear to me why it fails because the `string=` is passed the same to strings (and trying `gyy` in that situation works locally):
```
      (string= "(lorem (|ipsum (;; dolor
               dolor
               (sit)))
       (amet))" "(lorem (|ipsum (;; dolor
               dolor
               (sit)))
       (amet))")
```

I'm merging anyway because travis keeps failing before even testing anyway. I don't want to let that get in the way of merging your changes. Hopefully, I will have the time to switch the testing over to buttercup and fix this mess soon. If any other issues come up, we can address them later. Thanks again!
